{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b46d88e-13c7-4268-a7ab-132dafdaad00",
   "metadata": {},
   "outputs": [],
   "source": [
    "from pathlib import Path\n",
    "import gzip\n",
    "\n",
    "import bf\n",
    "import kbf\n",
    "import build_kmer_set\n",
    "import sparse_kbf\n",
    "import time\n",
    "import random\n",
    "\n",
    "from statistics import mean\n",
    "import math\n",
    "import pandas as pd\n",
    "import pickle"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f6831a35-2c4a-4d80-b3dd-95d2483fe368",
   "metadata": {
    "tags": []
   },
   "source": [
    "You can download the data we used from Zenodo (https://doi.org/10.5281/zenodo.5760011).\n",
    "You'll have to create a Data directory on the path of this notebook, and create 2 subdirectories: Genomes and Reads. Put the fasta files in the Genomes folder, fastq files in Reads folder."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "77fdcc16-180c-4805-be21-3734146692b5",
   "metadata": {},
   "outputs": [],
   "source": [
    "p = Path('./Data')\n",
    "g = p / 'Genomes' # directory of gzipped monkey, malaria, covid, and zebrafish genomes\n",
    "r = p / 'Reads' # directory of gzipped monkey, malaria, covid, and zebrafish reads"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "f5c6c5bb-7341-40f7-b340-88849cab8ed8",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_fasta(fh, limit):\n",
    "    count = 0\n",
    "    string = []\n",
    "    while limit >= count:\n",
    "        seq = fh.readline()\n",
    "        if len(seq) == 0:\n",
    "            break\n",
    "        seq = seq.decode(\"utf-8\").rstrip()\n",
    "        string.append(seq)\n",
    "        count += 1\n",
    "    return [\"\".join(string[1:])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "89eb5df2-aaa8-4e27-a5dc-ad971c699585",
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_fastq(fh, limit):\n",
    "    '''\n",
    "    Parse reads from a FASTQ filehandle. For each read, return a name,\n",
    "    nucleotide-string, quality-string triple.\n",
    "\n",
    "    Taken from Professor Langmead.\n",
    "    '''\n",
    "    reads = []\n",
    "    count = 0\n",
    "    while limit >= count:\n",
    "        first_line = fh.readline()  # ignore first line\n",
    "        if len(first_line) == 0:\n",
    "            break  # end of file\n",
    "        seq = fh.readline()\n",
    "        seq = seq.decode(\"utf-8\").rstrip()\n",
    "        fh.readline()  # ignore line starting with +\n",
    "        fh.readline()  # ignore quality\n",
    "        reads.append(seq)  # we only care about sequences\n",
    "        count += 1\n",
    "    return reads\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "822bd3a9-321a-44d7-82b7-1cf5e59c2f17",
   "metadata": {},
   "outputs": [],
   "source": [
    "def get_kmers(data, k):\n",
    "    kmers = set()\n",
    "    for read in data:\n",
    "        for i in range(len(read) - k + 1):  # for each k-mer\n",
    "            kmer = read[i:i + k]\n",
    "            kmers.add(kmer)\n",
    "    return kmers"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "6dd1b27c-a4a3-4b77-a039-3452f52b888e",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_KBF(kmers, k, num_hash, size, extend_len):\n",
    "    KBF = kbf.KmerBloomFilter(size, num_hash, range(num_hash), k, kmers, extend_len)\n",
    "    return KBF"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c8007180-09a5-4298-af11-73be60e8f297",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_KBF_Sparse(genome, k, num_hash, size, skip_len):\n",
    "    best_fit_set, edge_set = build_kmer_set.best_fit_kmers(genome, k, skip_len)\n",
    "    KBF_Sparse = sparse_kbf.KBFSparse(size, num_hash, \n",
    "                                     range(num_hash), k, skip_len, best_fit_set, edge_set)\n",
    "    return best_fit_set, edge_set, KBF_Sparse"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "008b7764-4790-4883-b98e-f55062c25c80",
   "metadata": {},
   "outputs": [],
   "source": [
    "def make_all_KBFs(file, k, skip_len, num_hash, alpha, extend_len):\n",
    "    fh = gzip.open(g / file, \"r\")\n",
    "    genome = parse_fasta(fh, 10000)\n",
    "    kmers = get_kmers(genome, k)\n",
    "    size = round(alpha * len(kmers))\n",
    "    \n",
    "    BF = make_KBF(kmers, k, num_hash, size, 0)\n",
    "    KBF = make_KBF(kmers, k, num_hash, size, extend_len)\n",
    "    best_fit_set, edge_set, KBF_Sparse = make_KBF_Sparse(genome, k, num_hash, size, skip_len)\n",
    "    \n",
    "    return {\n",
    "            \"genome\": genome, \"kmers\": kmers, \"KBF\": KBF,\n",
    "            \"best_fit_set\": best_fit_set, \"edge_set\": edge_set, \n",
    "            \"KBF_Sparse\": KBF_Sparse, \"BF\": BF, \"name\": file\n",
    "           }\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "id": "de78a168-83f4-49bf-866f-0c17faf7837a",
   "metadata": {},
   "outputs": [],
   "source": [
    "num_hash = 6\n",
    "k = 20\n",
    "alpha = 5\n",
    "skip_len = 3\n",
    "extend_len = 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "a04125fb-4452-449a-9bbd-e8090e8e42ca",
   "metadata": {},
   "outputs": [],
   "source": [
    "plas_Bloom = make_all_KBFs('plasmodium.fa.gz', k, skip_len, num_hash, alpha, extend_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "e88cb804-2145-4207-88fe-a86cff36c130",
   "metadata": {},
   "outputs": [],
   "source": [
    "covid_Bloom = make_all_KBFs('SARS-CoV-2.fa.gz', k, skip_len, num_hash, alpha, extend_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "6aaa9305-4b29-4cb8-a9c0-920bb394e18b",
   "metadata": {},
   "outputs": [],
   "source": [
    "zfish_Bloom = make_all_KBFs('Zebrafish.fa.gz', k, skip_len, num_hash, alpha, extend_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "747313cf-6d8b-41e9-b88f-f8620f389dcb",
   "metadata": {},
   "outputs": [],
   "source": [
    "monkey_Bloom = make_all_KBFs('monkey.fa.gz', k, skip_len, num_hash, alpha, extend_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "244ccc2c-e475-408d-a4ac-7bdeef8787e0",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 Data/Reads/sars-cov-2.fastq.gz\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/var/folders/bv/z4chszkx0gx32hlqmjdzsc5c0000gn/T/ipykernel_80867/3122238093.py:35: DeprecationWarning: Sampling from a set deprecated\n",
      "since Python 3.9 and will be removed in a subsequent version.\n",
      "  for x in random.sample(kmers, 100000):\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: monkey.fa.gz, KBF: BF\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [12210, 12103, 12173, 12172, 12130], PosixPath('Data/Reads/plasmodium.fastq.gz'): [12200, 12204, 12311, 12208, 12284], PosixPath('Data/Reads/zebrafish.fastq.gz'): [11544, 11508, 11624, 11558, 11763], PosixPath('Data/Reads/monkey.fastq.gz'): [12441, 12584, 12435, 12250, 12502]}, 'FPRs': [0.1184061216497123, 0.11491935483870967, 0.1151657013674239, 0.11617272984212863, 0.11742911077194955, 0.11515591300316462, 0.11486756821637192, 0.11809689070034907, 0.11788397412719456, 0.11596701346882814, 0.11604553001660366, 0.11667389615760963, 0.11795366213732639, 0.11479476088205935, 0.11536768824517885, 0.11508440733345435, 0.11757617144349153, 0.11580178218620217, 0.11734755121638925, 0.11687761157875613], 'Runtimes': [0.37125468254089355, 0.393878698348999, 0.42644619941711426, 0.6535229682922363, 0.33730006217956543, 0.409315824508667, 0.4317150115966797, 0.5131919384002686, 0.3460531234741211, 0.3898441791534424, 0.3811321258544922, 0.4541459083557129, 0.3267698287963867, 0.402601957321167, 0.4323408603668213, 0.4604461193084717, 0.3070211410522461, 0.37254929542541504, 0.37349605560302734, 0.41884708404541016]}\n",
      "1 Data/Reads/sars-cov-2.fastq.gz\n",
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: monkey.fa.gz, KBF: KBF\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [10594, 10624, 10466, 10549, 10599], PosixPath('Data/Reads/plasmodium.fastq.gz'): [10801, 10921, 10674, 10747, 10564], PosixPath('Data/Reads/zebrafish.fastq.gz'): [9876, 9933, 9902, 9840, 10161], PosixPath('Data/Reads/monkey.fastq.gz'): [10964, 11022, 10695, 10874, 10825]}, 'FPRs': [0.10206090310139804, 0.10065334435683894, 0.09837230008903829, 0.10136355837261175, 0.10247037557742518, 0.10191757067387185, 0.09902366804713603, 0.1019489498279151, 0.10084759379770225, 0.09941827053948601, 0.09872257845088879, 0.09882136874608972, 0.10170819148615672, 0.10047166958940558, 0.0981204173293721, 0.100482428695424, 0.10224636735185724, 0.09905408536401092, 0.1013134334330329, 0.1003783102143758], 'Runtimes': [0.8076841831207275, 0.8676009178161621, 0.8512511253356934, 0.9289536476135254, 0.7959742546081543, 0.8684959411621094, 0.8728761672973633, 0.9623610973358154, 0.8147437572479248, 0.9696049690246582, 0.921684980392456, 1.0744118690490723, 0.8524281978607178, 1.026191234588623, 0.9274609088897705, 1.0281198024749756, 0.8041641712188721, 0.8855719566345215, 0.8537321090698242, 0.954409122467041]}\n",
      "1 Data/Reads/sars-cov-2.fastq.gz\n",
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: monkey.fa.gz, KBF: KBF_Sparse\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [231, 238, 219, 219, 227], PosixPath('Data/Reads/plasmodium.fastq.gz'): [362, 339, 365, 341, 332], PosixPath('Data/Reads/zebrafish.fastq.gz'): [25, 31, 24, 21, 24], PosixPath('Data/Reads/monkey.fastq.gz'): [351, 389, 384, 379, 364]}, 'FPRs': [0.00015067199710709766, 0.00011087480218927336, 7.001750437609402e-05, 9.086044844679112e-05, 0.0001506629168340699, 0.00013104706605780183, 9.00297098042354e-05, 0.00016147914900488476, 0.0001606877435423613, 0.00016133911465160834, 4.001240384519201e-05, 0.00012110691722342208, 0.00017071529709482733, 0.00016120906801007557, 4.0012003601080326e-05, 0.00012114237257336685, 0.0001606732207951316, 0.00014110058455956461, 4.0015205778195714e-05, 0.0001311594495338795], 'Runtimes': [7.841922044754028, 7.824789047241211, 8.022415161132812, 8.431634902954102, 7.745975971221924, 7.948680877685547, 7.952022075653076, 8.28925108909607, 7.889333009719849, 7.749928951263428, 8.113046884536743, 8.092975854873657, 7.758435010910034, 7.968905925750732, 8.032985925674438, 8.266650199890137, 7.79744815826416, 8.062753200531006, 7.911250829696655, 8.047297954559326]}\n",
      "1 Data/Reads/sars-cov-2.fastq.gz\n",
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: plasmodium.fa.gz, KBF: BF\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [12027, 12023, 11979, 12032, 12009], PosixPath('Data/Reads/plasmodium.fastq.gz'): [13532, 13377, 13447, 13381, 13484], PosixPath('Data/Reads/zebrafish.fastq.gz'): [11561, 11607, 11680, 11496, 11827], PosixPath('Data/Reads/monkey.fastq.gz'): [11656, 11655, 11724, 11881, 11579]}, 'FPRs': [0.11631107360977178, 0.11801546339174605, 0.11518528893869057, 0.11591461767089975, 0.11586236005868993, 0.11720883779707308, 0.1157251328017927, 0.11587806733117169, 0.11566013282027067, 0.11732851985559567, 0.11652612309816043, 0.11658627384264356, 0.11605052403107007, 0.11613265306122449, 0.11459698476375314, 0.11813978623754053, 0.11599706641751309, 0.11741782792320404, 0.11789951779747494, 0.11520603598374927], 'Runtimes': [0.36019110679626465, 0.43775510787963867, 0.45006895065307617, 0.578557014465332, 0.3308429718017578, 0.3876650333404541, 0.37648582458496094, 0.42412590980529785, 0.3263978958129883, 0.3887200355529785, 0.4003300666809082, 0.42726588249206543, 0.31118011474609375, 0.37607288360595703, 0.3699009418487549, 0.428577184677124, 0.3114488124847412, 0.3725719451904297, 0.36535096168518066, 0.4141700267791748]}\n",
      "1 Data/Reads/sars-cov-2.fastq.gz\n",
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: plasmodium.fa.gz, KBF: KBF\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [10420, 10478, 10453, 10509, 10477], PosixPath('Data/Reads/plasmodium.fastq.gz'): [11861, 11952, 11898, 11958, 11801], PosixPath('Data/Reads/zebrafish.fastq.gz'): [10134, 9957, 10003, 10089, 10046], PosixPath('Data/Reads/monkey.fastq.gz'): [10231, 10103, 10207, 10022, 10178]}, 'FPRs': [0.0999789010459053, 0.10142016781020931, 0.10100737272790933, 0.10153732209700342, 0.10067909668086475, 0.10202749561457186, 0.09923671758550663, 0.10038227523817149, 0.10020197148283243, 0.10141261665561732, 0.09965185377858701, 0.10140503972939975, 0.10082792436147338, 0.10180471531610573, 0.10055720616628153, 0.09945453635590251, 0.10053351284550231, 0.10044162493498016, 0.10019005701710514, 0.10102486088314185], 'Runtimes': [0.8090827465057373, 0.915963888168335, 0.878950834274292, 0.9013240337371826, 0.7941627502441406, 0.9000260829925537, 0.8750228881835938, 1.0811400413513184, 0.8694078922271729, 0.9767611026763916, 0.9449760913848877, 0.9911041259765625, 0.8287041187286377, 0.9269287586212158, 0.8815810680389404, 0.9199488162994385, 0.8023688793182373, 0.917482852935791, 0.8830957412719727, 1.018517017364502]}\n",
      "1 Data/Reads/sars-cov-2.fastq.gz\n",
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: plasmodium.fa.gz, KBF: KBF_Sparse\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [307, 311, 311, 308, 318], PosixPath('Data/Reads/plasmodium.fastq.gz'): [886, 847, 917, 895, 944], PosixPath('Data/Reads/zebrafish.fastq.gz'): [37, 29, 36, 32, 36], PosixPath('Data/Reads/monkey.fastq.gz'): [82, 70, 75, 73, 65]}, 'FPRs': [0.0003415746591788143, 0.00038778675810270224, 4.001680705896477e-05, 0.00011009247768125225, 0.00031145136335322606, 0.0004790492401463648, 4.001640672675797e-05, 9.007295909686847e-05, 0.00029142799718621243, 0.0005407832173540396, 8.003601620729328e-05, 8.00640512409928e-05, 0.0003114826574494594, 0.0004792006525285481, 6.001920614596671e-05, 9.007295909686847e-05, 0.00035170577299904537, 0.0006939766905476293, 9.003871664815871e-05, 6.0045634682358595e-05], 'Runtimes': [7.721580982208252, 7.838573932647705, 7.910149812698364, 8.28108286857605, 7.894972085952759, 8.00826907157898, 7.864546060562134, 8.046983003616333, 7.702394962310791, 8.37879490852356, 8.005137920379639, 7.846832036972046, 7.711617946624756, 8.058192253112793, 7.841470956802368, 7.895013093948364, 7.578397989273071, 7.717328071594238, 7.73607611656189, 8.140040874481201]}\n",
      "1 Data/Reads/sars-cov-2.fastq.gz\n",
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: SARS-CoV-2.fa.gz, KBF: BF\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [25126, 25088, 25089, 25156, 25146], PosixPath('Data/Reads/plasmodium.fastq.gz'): [11835, 11463, 11585, 11623, 11621], PosixPath('Data/Reads/zebrafish.fastq.gz'): [11591, 11546, 11508, 11560, 11457], PosixPath('Data/Reads/monkey.fastq.gz'): [11512, 11499, 11600, 11789, 11660]}, 'FPRs': [0.11762418242884921, 0.11835, 0.11591, 0.11512, 0.11801827258170858, 0.11463, 0.11546, 0.11499, 0.1172609648605972, 0.11585, 0.11508, 0.116, 0.11742647579066531, 0.11622116221162211, 0.11559115591155912, 0.11789, 0.11714198098742717, 0.11621, 0.11457, 0.11659116591165912], 'Runtimes': [0.31890106201171875, 0.3830840587615967, 0.39344191551208496, 0.4415590763092041, 0.2949249744415283, 0.33621978759765625, 0.33704090118408203, 0.4434690475463867, 0.3681819438934326, 0.3803219795227051, 0.4165680408477783, 0.43707895278930664, 0.30396103858947754, 0.3417811393737793, 0.36066102981567383, 0.39282703399658203, 0.30182313919067383, 0.34885501861572266, 0.36357998847961426, 0.376816987991333]}\n",
      "1 Data/Reads/sars-cov-2.fastq.gz\n",
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: SARS-CoV-2.fa.gz, KBF: KBF\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [23745, 23670, 23578, 23655, 23669], PosixPath('Data/Reads/plasmodium.fastq.gz'): [10067, 10088, 9863, 10171, 10006], PosixPath('Data/Reads/zebrafish.fastq.gz'): [9997, 10070, 10007, 9866, 9987], PosixPath('Data/Reads/monkey.fastq.gz'): [9987, 10056, 9955, 10088, 10013]}, 'FPRs': [0.10186799208517855, 0.1006610066100661, 0.09997, 0.0998609986099861, 0.10083637648721876, 0.1008710087100871, 0.1007, 0.10056, 0.10043082138569108, 0.09863, 0.1000610006100061, 0.09955, 0.10039474459435574, 0.1017010170101701, 0.09864197283945679, 0.1008710087100871, 0.1007716322082818, 0.10006, 0.09987, 0.10013], 'Runtimes': [1.0852789878845215, 0.8890588283538818, 0.8881559371948242, 1.0176877975463867, 1.0761408805847168, 0.8751931190490723, 0.8504087924957275, 1.030634880065918, 1.07539701461792, 0.851240873336792, 0.8433501720428467, 1.0463542938232422, 1.0529038906097412, 0.8697819709777832, 0.842923641204834, 0.8774478435516357, 1.0722379684448242, 0.7949278354644775, 0.7819788455963135, 0.9180700778961182]}\n",
      "1 Data/Reads/sars-cov-2.fastq.gz\n",
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: SARS-CoV-2.fa.gz, KBF: KBF_Sparse\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [3885, 3886, 3962, 3937, 3912], PosixPath('Data/Reads/plasmodium.fastq.gz'): [3, 0, 5, 5, 3], PosixPath('Data/Reads/zebrafish.fastq.gz'): [4, 1, 1, 2, 2], PosixPath('Data/Reads/monkey.fastq.gz'): [2, 4, 1, 3, 6]}, 'FPRs': [1.1771630370806357e-05, 3e-05, 4e-05, 2e-05, 2.3532180256500763e-05, 0.0, 0.0, 3.000030000300003e-05, 1.1778146826378338e-05, 5e-05, 1e-05, 0.0, 1.1771214671641967e-05, 5e-05, 2e-05, 3e-05, 2.3553242103775585e-05, 2.000020000200002e-05, 2e-05, 5.000050000500005e-05], 'Runtimes': [7.086798906326294, 7.13959813117981, 6.938833236694336, 7.084367752075195, 7.050161123275757, 7.010969161987305, 6.937897205352783, 7.9653260707855225, 7.2435033321380615, 6.982064723968506, 7.1846020221710205, 7.0713982582092285, 6.969207048416138, 6.877220153808594, 6.8595969676971436, 7.043755054473877, 6.930617094039917, 7.09062385559082, 7.097610950469971, 7.219738960266113]}\n",
      "1 Data/Reads/sars-cov-2.fastq.gz\n",
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: Zebrafish.fa.gz, KBF: BF\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [11706, 11709, 11785, 11882, 11859], PosixPath('Data/Reads/plasmodium.fastq.gz'): [11842, 11867, 11841, 11846, 11842], PosixPath('Data/Reads/zebrafish.fastq.gz'): [12234, 11997, 12248, 12043, 11964], PosixPath('Data/Reads/monkey.fastq.gz'): [11673, 11750, 11688, 11562, 11713]}, 'FPRs': [0.11653875786714162, 0.1176965111391341, 0.11729978175381428, 0.11642958605926015, 0.11652457572846622, 0.11802615909614018, 0.11531656513259746, 0.11704086124784888, 0.11724972981627507, 0.1175805256941525, 0.1180880786315853, 0.11650893375217591, 0.11832626270711598, 0.11770122305182457, 0.11590342553875845, 0.11523955300779336, 0.1179725808065646, 0.1176081995435801, 0.11486024532475367, 0.11679438186510874], 'Runtimes': [0.34604907035827637, 0.4453108310699463, 0.4702482223510742, 0.531872034072876, 0.3867201805114746, 0.4417870044708252, 0.4919412136077881, 0.5172562599182129, 0.3498680591583252, 0.41023898124694824, 0.39500904083251953, 0.417949914932251, 0.3615751266479492, 0.3663010597229004, 0.37613487243652344, 0.4481947422027588, 0.30759501457214355, 0.363419771194458, 0.3689699172973633, 0.44449687004089355]}\n",
      "1 Data/Reads/sars-cov-2.fastq.gz\n",
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: Zebrafish.fa.gz, KBF: KBF\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [10138, 10189, 10165, 10216, 10137], PosixPath('Data/Reads/plasmodium.fastq.gz'): [10094, 10230, 10022, 10138, 10218], PosixPath('Data/Reads/zebrafish.fastq.gz'): [10538, 10620, 10620, 10571, 10584], PosixPath('Data/Reads/monkey.fastq.gz'): [10066, 10388, 10077, 10292, 10036]}, 'FPRs': [0.10082250995617283, 0.1002291810530319, 0.10051377954735118, 0.10035412040093632, 0.10131484149855907, 0.10167116981887321, 0.1014918171217178, 0.10348553363479931, 0.10115563559957977, 0.0995446584938704, 0.10099475965842226, 0.10041016406562625, 0.10170186795265586, 0.10075052536775743, 0.10110767127693793, 0.10253408965855318, 0.10089347354097672, 0.10146116893514812, 0.10101243678553834, 0.10004501575551443], 'Runtimes': [0.858799934387207, 0.8711340427398682, 0.8802859783172607, 0.8974719047546387, 0.7975239753723145, 0.8567366600036621, 0.9150640964508057, 0.9294369220733643, 0.826948881149292, 0.871492862701416, 0.8803620338439941, 0.9763970375061035, 0.795363187789917, 0.8602321147918701, 0.8779401779174805, 1.179595947265625, 0.8679318428039551, 1.122281789779663, 0.9901978969573975, 1.126723051071167]}\n",
      "1 Data/Reads/sars-cov-2.fastq.gz\n",
      "1 Data/Reads/plasmodium.fastq.gz\n",
      "1 Data/Reads/zebrafish.fastq.gz\n",
      "1 Data/Reads/monkey.fastq.gz\n",
      "2 Data/Reads/sars-cov-2.fastq.gz\n",
      "2 Data/Reads/plasmodium.fastq.gz\n",
      "2 Data/Reads/zebrafish.fastq.gz\n",
      "2 Data/Reads/monkey.fastq.gz\n",
      "3 Data/Reads/sars-cov-2.fastq.gz\n",
      "3 Data/Reads/plasmodium.fastq.gz\n",
      "3 Data/Reads/zebrafish.fastq.gz\n",
      "3 Data/Reads/monkey.fastq.gz\n",
      "4 Data/Reads/sars-cov-2.fastq.gz\n",
      "4 Data/Reads/plasmodium.fastq.gz\n",
      "4 Data/Reads/zebrafish.fastq.gz\n",
      "4 Data/Reads/monkey.fastq.gz\n",
      "5 Data/Reads/sars-cov-2.fastq.gz\n",
      "5 Data/Reads/plasmodium.fastq.gz\n",
      "5 Data/Reads/zebrafish.fastq.gz\n",
      "5 Data/Reads/monkey.fastq.gz\n",
      "Bloom: Zebrafish.fa.gz, KBF: KBF_Sparse\n",
      "{'classification': {PosixPath('Data/Reads/sars-cov-2.fastq.gz'): [49, 44, 46, 53, 52], PosixPath('Data/Reads/plasmodium.fastq.gz'): [50, 57, 50, 62, 58], PosixPath('Data/Reads/zebrafish.fastq.gz'): [210, 202, 198, 206, 208], PosixPath('Data/Reads/monkey.fastq.gz'): [21, 39, 27, 31, 23]}, 'FPRs': [0.00011006934368652251, 8.005924384044192e-05, 5.0261359067149175e-05, 6.001740504746376e-05, 7.004062356166576e-05, 8.006485253054975e-05, 9.048045119584996e-05, 5.002801568878572e-05, 0.00010005603137757145, 2.0016814123864046e-05, 8.040766687103615e-05, 6.002400960384154e-05, 0.00011007264794764545, 0.0001101001911739683, 0.00013069792692980517, 4.001600640256102e-05, 0.00010006303971502046, 3.0029729432137816e-05, 0.00011057610148875642, 5.001700578196587e-05], 'Runtimes': [8.2422935962677, 8.067845106124878, 7.913371801376343, 7.946472883224487, 7.830396890640259, 8.024813175201416, 8.067209959030151, 8.340571165084839, 7.798586130142212, 7.978209018707275, 7.97692084312439, 8.836294889450073, 7.906548261642456, 8.11381220817566, 7.93546986579895, 7.897680044174194, 7.735925197601318, 8.125892162322998, 8.131412029266357, 8.098428010940552]}\n"
     ]
    }
   ],
   "source": [
    "KMERS = {}\n",
    "for file in r.iterdir():\n",
    "    fh = gzip.open(file, \"r\")\n",
    "    reads = parse_fastq(fh, 10000)\n",
    "    kmers = get_kmers(reads, k)\n",
    "    KMERS[file] = kmers\n",
    "\n",
    "\n",
    "All_Classifications = {}\n",
    "for Bloom in [monkey_Bloom, plas_Bloom, covid_Bloom, zfish_Bloom]:\n",
    "    All_Classifications[Bloom[\"name\"]] = {}\n",
    "    \n",
    "    for KBF_Implementation in [\"BF\", \"KBF\", \"KBF_Sparse\"]:\n",
    "        All_Classifications[Bloom[\"name\"]][KBF_Implementation] = {}\n",
    "        \n",
    "        classification = {}\n",
    "        for file in r.iterdir():\n",
    "            classification[file] = [] \n",
    "            \n",
    "        FPRs = []\n",
    "        Runtimes = []\n",
    "        \n",
    "        for i in range(5): \n",
    "            for file in r.iterdir():\n",
    "                print(f\"{i + 1} {file}\")\n",
    "                kmers = KMERS[file]\n",
    "\n",
    "                tp = 0\n",
    "                tn = 0\n",
    "                fp = 0\n",
    "                fn = 0\n",
    "                pos = 0\n",
    "\n",
    "                start = time.time()\n",
    "                for x in random.sample(kmers, 100000):\n",
    "                    truth = x in Bloom[\"kmers\"]\n",
    "                    if x in Bloom[KBF_Implementation]:\n",
    "                        pos += 1\n",
    "                        if truth:\n",
    "                            tp += 1\n",
    "                        else:\n",
    "                            fp += 1\n",
    "                    else:\n",
    "                        if truth:\n",
    "                            fn += 1\n",
    "                        else:\n",
    "                            tn += 1\n",
    "                end = time.time()\n",
    "\n",
    "                classification[file].append(pos)\n",
    "                FPRs.append(fp / (fp + tn))\n",
    "                Runtimes.append(end - start)\n",
    "                \n",
    "        All_Classifications[Bloom[\"name\"]][KBF_Implementation][\"classification\"] = classification\n",
    "        All_Classifications[Bloom[\"name\"]][KBF_Implementation][\"FPRs\"] = FPRs\n",
    "        All_Classifications[Bloom[\"name\"]][KBF_Implementation][\"Runtimes\"] = Runtimes\n",
    "        \n",
    "        print(\"Bloom: \" + Bloom[\"name\"] + \", KBF: \" + KBF_Implementation)\n",
    "        print(All_Classifications[Bloom[\"name\"]][KBF_Implementation])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "id": "db7a6207-491b-4582-bf50-567d895987a7",
   "metadata": {},
   "outputs": [],
   "source": [
    "filehandler = open(p / 'Results/All_Classifications.pkl', 'wb') \n",
    "pickle.dump(All_Classifications, filehandler)\n",
    "filehandler.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 131,
   "id": "5a716a8a-8f21-4659-b96b-550d04f367a5",
   "metadata": {
    "tags": []
   },
   "outputs": [],
   "source": [
    "Results = pickle.load(open(p / 'Results/All_Classifications.pkl', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 132,
   "id": "9285244c-7458-4433-a2f9-99f4e121f558",
   "metadata": {},
   "outputs": [],
   "source": [
    "Genomes = [\"M. mulatta\", \"P. falciparum\", \"SARS-CoV-2\", \"D. rerio\"]\n",
    "BF_FPRs = []\n",
    "KBF_FPRs = []\n",
    "Sparse_FPRs = []\n",
    "BF_Runtimes = []\n",
    "KBF_Runtimes = []\n",
    "Sparse_Runtimes = []\n",
    "\n",
    "\n",
    "Kmer_lens = []\n",
    "Bestfit_lens = []\n",
    "Sizes = []\n",
    "\n",
    "for Bloom in [monkey_Bloom, plas_Bloom, covid_Bloom, zfish_Bloom]:\n",
    "    Kmer_lens.append(len(Bloom[\"kmers\"]))\n",
    "    Bestfit_lens.append(len(Bloom[\"best_fit_set\"]))\n",
    "    Sizes.append(round(len(Bloom[\"kmers\"]) * alpha))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "id": "169e0492-479b-48a2-84af-1ba2e3877541",
   "metadata": {
    "scrolled": true,
    "tags": []
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Classifying species of Data/Reads/sars-cov-2.fastq.gz...\n",
      "Via BF implementation\n",
      "great! we classified correctly: sars-cov-2\n",
      "Via KBF implementation\n",
      "great! we classified correctly: sars-cov-2\n",
      "Via KBF_Sparse implementation\n",
      "great! we classified correctly: sars-cov-2\n",
      "\n",
      "\n",
      "Classifying species of Data/Reads/plasmodium.fastq.gz...\n",
      "Via BF implementation\n",
      "great! we classified correctly: plasmodium\n",
      "Via KBF implementation\n",
      "great! we classified correctly: plasmodium\n",
      "Via KBF_Sparse implementation\n",
      "great! we classified correctly: plasmodium\n",
      "\n",
      "\n",
      "Classifying species of Data/Reads/zebrafish.fastq.gz...\n",
      "Via BF implementation\n",
      "great! we classified correctly: zebrafish\n",
      "Via KBF implementation\n",
      "great! we classified correctly: zebrafish\n",
      "Via KBF_Sparse implementation\n",
      "great! we classified correctly: zebrafish\n",
      "\n",
      "\n",
      "Classifying species of Data/Reads/monkey.fastq.gz...\n",
      "Via BF implementation\n",
      "great! we classified correctly: monkey\n",
      "Via KBF implementation\n",
      "great! we classified correctly: monkey\n",
      "Via KBF_Sparse implementation\n",
      "great! we classified correctly: monkey\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for x in r.iterdir():\n",
    "    print(f\"Classifying species of {x}...\")\n",
    "    for kbf_type in [\"BF\", \"KBF\", \"KBF_Sparse\"]:\n",
    "        print(f\"Via {kbf_type} implementation\")\n",
    "        max_hit = -math.inf\n",
    "        max_word = \"\"\n",
    "        for genome_type in Results:\n",
    "            hits = mean(Results[genome_type][kbf_type][\"classification\"][x])\n",
    "            if hits > max_hit:\n",
    "                max_hit = hits\n",
    "                max_word = genome_type.split(\".\")[0].lower()\n",
    "            # print(f\"genome: {genome_type}, # of hits: {hits}\")\n",
    "        if max_word == str(x).split(\"/\")[-1].split(\".\")[0].lower():\n",
    "            print(f\"great! we classified correctly: {max_word}\")\n",
    "        else:\n",
    "            print(f\"WARNING: WRONG CLASSIFICATION: {max_word}\")\n",
    "    print('\\n')\n",
    "            \n",
    "for key in [\"monkey\", \"plasmodium\", \"SARS-CoV-2\", \"Zebrafish\"]:\n",
    "    genome_type = key + \".fa.gz\"\n",
    "    for kbf_type in Results[genome_type]:\n",
    "        core_result = Results[genome_type][kbf_type]\n",
    "        \n",
    "        if kbf_type == \"BF\":\n",
    "            BF_FPRs.append(mean(core_result[\"FPRs\"]))\n",
    "            BF_Runtimes.append(mean(core_result[\"Runtimes\"]))\n",
    "        elif kbf_type == \"KBF\":\n",
    "            KBF_FPRs.append(mean(core_result[\"FPRs\"]))\n",
    "            KBF_Runtimes.append(mean(core_result[\"Runtimes\"]))\n",
    "        elif kbf_type == \"KBF_Sparse\":\n",
    "            Sparse_FPRs.append(mean(core_result[\"FPRs\"]))\n",
    "            Sparse_Runtimes.append(mean(core_result[\"Runtimes\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 135,
   "id": "70416492-da8f-4a74-84e4-817746b73f53",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>KBF</th>\n",
       "      <th>2-KBF</th>\n",
       "      <th>Sparse KBF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M. mulatta</td>\n",
       "      <td>0.116</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P. falciparum</td>\n",
       "      <td>0.116</td>\n",
       "      <td>0.101</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>0.116</td>\n",
       "      <td>0.100</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D. rerio</td>\n",
       "      <td>0.117</td>\n",
       "      <td>0.101</td>\n",
       "      <td>0.000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Genome   KBF  2-KBF  Sparse KBF\n",
       "0     M. mulatta 0.116  0.100       0.000\n",
       "1  P. falciparum 0.116  0.101       0.000\n",
       "2     SARS-CoV-2 0.116  0.100       0.000\n",
       "3       D. rerio 0.117  0.101       0.000"
      ]
     },
     "execution_count": 135,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "pd.options.display.float_format = '{:,.3f}'.format\n",
    "FPR_Dataset = list(zip(Genomes,BF_FPRs,KBF_FPRs,Sparse_FPRs))\n",
    "df = pd.DataFrame(data = FPR_Dataset, columns=['Genome', 'KBF', '2-KBF', 'Sparse KBF'])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 136,
   "id": "f73daead-b6d3-46be-b111-7a1152f5a098",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>KBF</th>\n",
       "      <th>2-KBF</th>\n",
       "      <th>Sparse KBF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M. mulatta</td>\n",
       "      <td>0.410</td>\n",
       "      <td>0.903</td>\n",
       "      <td>7.987</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P. falciparum</td>\n",
       "      <td>0.392</td>\n",
       "      <td>0.906</td>\n",
       "      <td>7.909</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>0.367</td>\n",
       "      <td>0.937</td>\n",
       "      <td>7.089</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D. rerio</td>\n",
       "      <td>0.412</td>\n",
       "      <td>0.919</td>\n",
       "      <td>8.048</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Genome   KBF  2-KBF  Sparse KBF\n",
       "0     M. mulatta 0.410  0.903       7.987\n",
       "1  P. falciparum 0.392  0.906       7.909\n",
       "2     SARS-CoV-2 0.367  0.937       7.089\n",
       "3       D. rerio 0.412  0.919       8.048"
      ]
     },
     "execution_count": 136,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Runtime_Dataset = list(zip(Genomes,BF_Runtimes,KBF_Runtimes,Sparse_Runtimes))\n",
    "df = pd.DataFrame(data = Runtime_Dataset, columns=['Genome', 'KBF', '2-KBF', 'Sparse KBF'])\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 137,
   "id": "e26db906-29f5-43bc-90b4-2eebb67e7cf6",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th># of k-mers</th>\n",
       "      <th># of k-mers (sparse)</th>\n",
       "      <th>KBF Size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M. mulatta</td>\n",
       "      <td>649,964</td>\n",
       "      <td>168,543</td>\n",
       "      <td>3,249,820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P. falciparum</td>\n",
       "      <td>571,486</td>\n",
       "      <td>149,881</td>\n",
       "      <td>2,857,430</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>29,871</td>\n",
       "      <td>7,469</td>\n",
       "      <td>149,355</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D. rerio</td>\n",
       "      <td>629,173</td>\n",
       "      <td>163,419</td>\n",
       "      <td>3,145,865</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Genome # of k-mers # of k-mers (sparse)   KBF Size\n",
       "0     M. mulatta     649,964              168,543  3,249,820\n",
       "1  P. falciparum     571,486              149,881  2,857,430\n",
       "2     SARS-CoV-2      29,871                7,469    149,355\n",
       "3       D. rerio     629,173              163,419  3,145,865"
      ]
     },
     "execution_count": 137,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "Size_Dataset = list(zip(Genomes,Kmer_lens,Bestfit_lens,Sizes))\n",
    "df = pd.DataFrame(data = Size_Dataset, \n",
    "                  columns=['Genome', '# of k-mers', '# of k-mers (sparse)', 'KBF Size'])\n",
    "for key in df:\n",
    "    if key != 'Genome':\n",
    "        df[key] = df[key].apply('{:,}'.format)\n",
    "df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "dd479d4d-478d-4d3d-944f-42f2d1d2cb09",
   "metadata": {},
   "source": [
    "We want to see if we can use a smaller sparse KBF to achieve similar FPR as a bigger two-sided KBF.\n",
    "Change alpha to 2."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 138,
   "id": "90198e95-525c-4484-a274-a77f4094553d",
   "metadata": {},
   "outputs": [],
   "source": [
    "alpha = 2 # decreased from 5 to 2, size of KBF will be reduced by a factor of 2.5"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62cd0569-9a2f-4112-8547-1518e23e538c",
   "metadata": {},
   "outputs": [],
   "source": [
    "plas_Bloom = make_all_KBFs('plasmodium.fa.gz', k, skip_len, num_hash, alpha, extend_len)\n",
    "covid_Bloom = make_all_KBFs('SARS-CoV-2.fa.gz', k, skip_len, num_hash, alpha, extend_len)\n",
    "zfish_Bloom = make_all_KBFs('Zebrafish.fa.gz', k, skip_len, num_hash, alpha, extend_len)\n",
    "monkey_Bloom = make_all_KBFs('monkey.fa.gz', k, skip_len, num_hash, alpha, extend_len)\n",
    "\n",
    "KMERS = {}\n",
    "for file in r.iterdir():\n",
    "    fh = gzip.open(file, \"r\")\n",
    "    reads = parse_fastq(fh, 10000)\n",
    "    kmers = get_kmers(reads, k)\n",
    "    KMERS[file] = kmers\n",
    "\n",
    "\n",
    "All_Classifications = {}\n",
    "for Bloom in [monkey_Bloom, plas_Bloom, covid_Bloom, zfish_Bloom]:\n",
    "    All_Classifications[Bloom[\"name\"]] = {}\n",
    "    \n",
    "    for KBF_Implementation in [\"KBF_Sparse\"]:\n",
    "        All_Classifications[Bloom[\"name\"]][KBF_Implementation] = {}\n",
    "        \n",
    "        classification = {}\n",
    "        for file in r.iterdir():\n",
    "            classification[file] = [] \n",
    "            \n",
    "        FPRs = []\n",
    "        Runtimes = []\n",
    "        \n",
    "        for i in range(5): \n",
    "            for file in r.iterdir():\n",
    "                print(f\"{i + 1} {file}\")\n",
    "                kmers = KMERS[file]\n",
    "\n",
    "                tp = 0\n",
    "                tn = 0\n",
    "                fp = 0\n",
    "                fn = 0\n",
    "                pos = 0\n",
    "\n",
    "                start = time.time()\n",
    "                for x in random.sample(kmers, 100000):\n",
    "                    truth = x in Bloom[\"kmers\"]\n",
    "                    if x in Bloom[KBF_Implementation]:\n",
    "                        pos += 1\n",
    "                        if truth:\n",
    "                            tp += 1\n",
    "                        else:\n",
    "                            fp += 1\n",
    "                    else:\n",
    "                        if truth:\n",
    "                            fn += 1\n",
    "                        else:\n",
    "                            tn += 1\n",
    "                end = time.time()\n",
    "\n",
    "                classification[file].append(pos)\n",
    "                FPRs.append(fp / (fp + tn))\n",
    "                Runtimes.append(end - start)\n",
    "                \n",
    "        All_Classifications[Bloom[\"name\"]][KBF_Implementation][\"classification\"] = classification\n",
    "        All_Classifications[Bloom[\"name\"]][KBF_Implementation][\"FPRs\"] = FPRs\n",
    "        All_Classifications[Bloom[\"name\"]][KBF_Implementation][\"Runtimes\"] = Runtimes\n",
    "        \n",
    "        print(\"Bloom: \" + Bloom[\"name\"] + \", KBF: \" + KBF_Implementation)\n",
    "        print(All_Classifications[Bloom[\"name\"]][KBF_Implementation])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "5bf09214-7e98-46a4-be05-7bae0f3da4b2",
   "metadata": {},
   "outputs": [],
   "source": [
    "filehandler = open(p / 'Results/Sparse_Classifications_Smaller.pkl', 'wb') \n",
    "pickle.dump(All_Classifications, filehandler)\n",
    "filehandler.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce9a55c6-e531-4172-b19f-45eb516dd7aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "Results = pickle.load(open(p / 'Results/Sparse_Classifications_Smaller.pkl', 'rb'))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 140,
   "id": "5cc2289e-a323-4cc4-8c8f-182c8e86e81f",
   "metadata": {},
   "outputs": [],
   "source": [
    "Genomes = [\"M. mulatta\", \"P. falciparum\", \"SARS-CoV-2\", \"D. rerio\"]\n",
    "Sparse_FPRs = []\n",
    "Sizes = []\n",
    "\n",
    "for Bloom in [monkey_Bloom, plas_Bloom, covid_Bloom, zfish_Bloom]:\n",
    "    Sizes.append(round(len(Bloom[\"kmers\"]) * alpha))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 141,
   "id": "dfd9dd46-12b3-4426-ad4d-d0499171fc1d",
   "metadata": {},
   "outputs": [],
   "source": [
    "for key in [\"monkey\", \"plasmodium\", \"SARS-CoV-2\", \"Zebrafish\"]:\n",
    "    genome_type = key + \".fa.gz\"\n",
    "    core_result = Results[genome_type][\"KBF_Sparse\"]\n",
    "    Sparse_FPRs.append(mean(core_result[\"FPRs\"]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 142,
   "id": "c9e5f3d3-2913-4a19-be97-fdd8b2766ca7",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genome</th>\n",
       "      <th>FPR</th>\n",
       "      <th>Size</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>M. mulatta</td>\n",
       "      <td>0.129</td>\n",
       "      <td>1,299,928</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>P. falciparum</td>\n",
       "      <td>0.140</td>\n",
       "      <td>1,142,972</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SARS-CoV-2</td>\n",
       "      <td>0.101</td>\n",
       "      <td>59,742</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>D. rerio</td>\n",
       "      <td>0.131</td>\n",
       "      <td>1,258,346</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          Genome   FPR       Size\n",
       "0     M. mulatta 0.129  1,299,928\n",
       "1  P. falciparum 0.140  1,142,972\n",
       "2     SARS-CoV-2 0.101     59,742\n",
       "3       D. rerio 0.131  1,258,346"
      ]
     },
     "execution_count": 142,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "FPR_Dataset = list(zip(Genomes,Sparse_FPRs,Sizes))\n",
    "df = pd.DataFrame(data = FPR_Dataset, columns=['Genome', 'FPR', 'Size'])\n",
    "df['Size'] = df['Size'].apply('{:,}'.format)\n",
    "df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a1c04f50-5b0f-466b-897f-3f653b04f9f4",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
